File: iso27001-symfony/src/Infrastructure/Encryption/FieldEncryptor.php
<?php

declare(strict_types=1);

namespace App\Infrastructure\Encryption;

/**
 * A.10: AES-256-GCM field-level encryption for PII at rest.
 * Used for encrypting email and phone fields in the database.
 * Key is loaded from environment / secrets manager — never hardcoded.
 */
final class FieldEncryptor
{
    private const CIPHER = 'aes-256-gcm';
    private const TAG_LENGTH = 16;

    public function __construct(private readonly string $key)
    {
        if (strlen($key) !== 32) {
            throw new \InvalidArgumentException(
                'Encryption key must be exactly 32 bytes for AES-256.'
            );
        }
    }

    /**
     * Encrypts plaintext.
     * Returns base64-encoded: IV (12 bytes) + TAG (16 bytes) + ciphertext
     */
    public function encrypt(string $plaintext): string
    {
        $iv = random_bytes(12); // 96-bit IV for GCM
        $tag = '';

        $ciphertext = openssl_encrypt(
            $plaintext,
            self::CIPHER,
            $this->key,
            OPENSSL_RAW_DATA,
            $iv,
            $tag,
            '',
            self::TAG_LENGTH,
        );

        if ($ciphertext === false) {
            throw new \RuntimeException('Encryption failed.');
        }

        return base64_encode($iv . $tag . $ciphertext);
    }

    /**
     * Decrypts a value produced by encrypt().
     */
    public function decrypt(string $encoded): string
    {
        $raw = base64_decode($encoded, strict: true);
        if ($raw === false) {
            throw new \InvalidArgumentException('Invalid base64 ciphertext.');
        }

        $iv = substr($raw, 0, 12);
        $tag = substr($raw, 12, self::TAG_LENGTH);
        $ciphertext = substr($raw, 12 + self::TAG_LENGTH);

        $plaintext = openssl_decrypt(
            $ciphertext,
            self::CIPHER,
            $this->key,
            OPENSSL_RAW_DATA,
            $iv,
            $tag,
        );

        if ($plaintext === false) {
            throw new \RuntimeException('Decryption failed — data may be tampered.');
        }

        return $plaintext;
    }
}

File: iso27001-phoenix/lib/iso27001_phoenix/infrastructure/encryption/field_encryptor.ex
defmodule Iso27001Phoenix.Infrastructure.Encryption.FieldEncryptor do
  @moduledoc """
  A.10: AES-256-GCM field-level encryption for PII at rest.

  Key:         32 bytes from ENCRYPTION_KEY env var
  IV:          12 random bytes per encrypt call
  Auth tag:    16 bytes (GCM default)
  Wire format: Base64(IV ++ TAG ++ ciphertext)
  """

  @iv_length  12
  @tag_length 16

  @doc "Encrypts plaintext string. Returns Base64-encoded ciphertext."
  def encrypt(plaintext) when is_binary(plaintext) do
    key = encryption_key!()
    iv  = :crypto.strong_rand_bytes(@iv_length)

    {ciphertext, tag} =
      :crypto.crypto_one_time_aead(:aes_256_gcm, key, iv, plaintext, "", true)

    Base.encode64(iv <> tag <> ciphertext)
  end

  @doc "Decrypts Base64-encoded ciphertext. Returns {:ok, plaintext} or {:error, reason}."
  def decrypt(encoded) when is_binary(encoded) do
    key = encryption_key!()
    raw = Base.decode64!(encoded)

    <<iv::binary-size(@iv_length), tag::binary-size(@tag_length), ciphertext::binary>> = raw

    case :crypto.crypto_one_time_aead(:aes_256_gcm, key, iv, ciphertext, "", tag, false) do
      plaintext when is_binary(plaintext) -> {:ok, plaintext}
      :error -> {:error, :decryption_failed}
    end
  rescue
    _ -> {:error, :decryption_failed}
  end

  defp encryption_key! do
    key = Application.get_env(:iso27001_phoenix, :encryption_key, "")
    unless byte_size(key) == 32 do
      raise "ENCRYPTION_KEY must be exactly 32 bytes, got #{byte_size(key)}"
    end
    key
  end
end
